### ISSUE TITLE
Che wont start because of Java but i have the right version.

### ISSUE THREAD

**Issue by griffinwebnet** [eclipse-che/che#258]
I have installed Che =by installing the prerequisites noted here:

https://eclipse-che.readme.io/docs/install-pre-reqs

then followed the readme instructions on github. everything seems to work until i try to start the server where i receive the error: 

`Che requires Java version 1.8 or higher. We found a lower version.`

the only problem is the java version im running is:

`Java version "1.8.0_72"`
`Java(TM) SE Runtime Environment (build 1.8.0_72-b15)`
`Java HotSpot(TM) 64-Bit Server VM (build 25.72-b15, mixed mode)`

This should be enough to satisfy the Java 1.8 Requirement should it not?

**Comment by TylerJewell** [comment]
Yes, absolutely.

Can you tell us:
- Which version of che you are running (will be on your command line)
- Which OS you have
- Which directory you installed Java into
- Is JAVA_HOME variable already set and if so, what is its value?

Thanks - we can help debug from here.

**Comment by griffinwebnet** [comment]
sure! where exactly am i looking for that version number? i dont see it in the output from the server start command. I can tell you it was done via a git checkout from master this morning.

Im currently running `Ubuntu Server 14.04_3 x64`
java is technically installed to `/usr/lib/jvm/java-8-oracle/jre/bin/java` but it is linked to `/usr/lib/java`
JAVA_HOME is pointed to `/usr/lib/java`

**Comment by griffinwebnet** [comment]
update: exporting the java_home variable with the full path of java 8 doesnt change the outcome when starting che

**Comment by TylerJewell** [comment]
Ok.  Something is failing with the version check.  It is happening in the launch script.  If you give me a moment I will give you a work around.

**Comment by griffinwebnet** [comment]
alright thanks!

**Comment by TylerJewell** [comment]
@griffinwebnet:

First, the quickest workaround to just avoid this issue.  If you can open the che.sh file where Eclipse Che is installed and find the following code.  Just comment it out with # symbols. It is around line 465.  This should let you continue on with the script.  

```
  # Che requires Java version 1.8 or higher.
  JAVA_VERSION=$("${JAVA_HOME}/bin/java" -version 2>&1 | awk -F '"' '/version/ {print $2}')
  if [[  -z "${JAVA_VERSION}" || "${JAVA_VERSION}" < "1.8" ]]; then
      error_exit "Che requires Java version 1.8 or higher. We found a lower version."
      return
  fi
```

Second, I will create an issue to add in more skip-check parameters on the command line so that users can skip all of the checks so that you can enact these work arounds while we diagnose particular issues.

Third, if you could be so kind, we'd like your help in debugging why this failure occured. Could you (on your terminal line) tell us the output of this:

```
java -version 2>&1 | awk -F '"' '/version/ {print $2}'
```

This should return "1.8".

**Comment by griffinwebnet** [comment]
i am more than happy to help in any way i can just let me know what you might need from me.

This seems to bypass that issue but now i recieve:

`Docker is natively installed and reachable...`
`############## HOW TO CONNECT YOUR CHE CLIENT ###############`
`After Che server has booted, you can connect your clients by:`
`1. Open browser to http://localhost:8080, or:`
`2. Open native chromium app.`
`#############################################################`
`./che.sh: line 481: /srv/che/assembly-main/src/assembly/tomcat/bin/catalina.sh:` 

i can confirm that catalina.sh is not present in this directory nor is it present on the git repo:
https://github.com/codenvy/che/tree/master/assembly-main/src/assembly/tomcat/bin

Additionally there is nothing accessible via port 8080 i get:
`ERR_CONNECTION_REFUSED`

**Comment by TylerJewell** [comment]
These are unusual issues.
- Could you download the nightly ZIP file from eclipse.org/che download page, extract it to a directory.  And does running this cause the same issue?
- Did you set CHE_HOME before you ran the script?  If so, you do not need to. It could be that you set it to the wrong directory, which should be like /srv/che/assembly-main/target/eclipse-che<version>/eclipse-che<version>/.

**Comment by griffinwebnet** [comment]
i will try the nightly and no i did not. by the way i do not have that directory at all `assembly-main/target/`

Its not in my files, nor is it in the git repo:
https://github.com/codenvy/che/tree/master/assembly-main

**Comment by TylerJewell** [comment]
Ahhh, well then that says it.  You checked out the source code but you didn't compile it. :)

So, you can download the binaries from eclipse.org/che (preferred), or:
You can compile the source code: https://eclipse-che.readme.io/docs/contribute#build

**Comment by griffinwebnet** [comment]
here is exactly what i did command by command:

`add-apt-repository ppa:webupd8team/java`
`apt-key adv --keyserver hkp://pgp.mit.edu:80 --recv-keys 58118E89F3A912897C070ADBF76221572C52609D`
`echo "deb https://apt.dockerproject.org/repo ubuntu-trusty main" | sudo tee /etc/apt/sources.list.d/docker.list`
`apt-get update`
`apt-get install oracle-java8-installer git maven docker-engine`
`cd /srv`
`git clone https://github.com/codenvy/che.git`
`mvn clean install`
`useradd che`
`passwd che`
`usermod -aG docker che`
`su che`
`export JAVA_HOME /usr/lib/java`
`cd ./che/assembly-main/src/assembly/bin/`
`./che.sh start`

**Comment by TylerJewell** [comment]
That is so odd.

After you mvn clean install Che, there should be an assembly-main/target directory.  If you don't have that, then the compilation failed.  

Can you go into the che/assembly-main directory and just run mvn clean install again and let us know the output?

**Comment by griffinwebnet** [comment]
sure!

okay out of the several hundred lines of output i managed to find this chunk:

`[INFO] --- maven-enforcer-plugin:1.4:enforce (enforce-maven-version) @ che-parent ---`
`[WARNING] Rule 0: org.apache.maven.plugins.enforcer.RequireMavenVersion failed with message:`
`To build this project Maven 3.3.1 (or upper) is required. Please install it.`
`[INFO] ------------------------------------------------------------------------`
`[INFO] Reactor Summary:`
`[INFO]`
`[INFO] Che IDE :: Parent ................................. FAILURE [13.000s]`
`[INFO] Che Machine Plugins :: Packaging .................. SKIPPED`
`[INFO] Che Assembly Machine Plugins :: Server ............ SKIPPED`
`[INFO] Che IDE :: Compiling GWT Application .............. SKIPPED`
`[INFO] Che IDE :: Assemblies Tomcat ...................... SKIPPED`
`[INFO] ------------------------------------------------------------------------`
`[INFO] BUILD FAILURE`
`[INFO] ------------------------------------------------------------------------`
`[INFO] Total time: 23.178s`
`[INFO] Finished at: Tue Feb 02 19:14:36 EST 2016`
`[INFO] Final Memory: 15M/37M`
`[INFO] ------------------------------------------------------------------------`
`[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:1.4:enforce (enforce-maven-version) on project che-parent: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed. -> [Help 1]`
`[ERROR]`
`[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.`
`[ERROR] Re-run Maven using the -X switch to enable full debug logging.`
`[ERROR]`
`[ERROR] For more information about the errors and possible solutions, please read the following articles:`
`[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException`

**Comment by TylerJewell** [comment]
The error is there :)  It says that maven 3.3.1 is required.  You have an older version of maven on your computer.  You can download and install a newer version of maven if you'd like.

Maven doesn't do a very good job of making it obvious, but that is the first error you got.

This basic issue would explain all of the other problems we were having with JAVA_HOME.

BTW, if you just install the binaries, then you won't have this maven requirement.

**Comment by griffinwebnet** [comment]
now errors aside everything right up to mvn clean install is right, correct?

if so id like to put together a small writeup for installing this standalone on an ubuntu server as there isn't a lot of documentation for doing that and much of my figuring this out has been hit and miss. seems most instructions are targeted towards Windows and Mac with little tid bits of hints about linux installs and how to set up its prerequisites (not a big problem for experienced linux admins, but you need all the information first before you can get to that point and its not all in one place like it is for other platforms)

Perhaps alongside a write up i can build an installer script to install the binaries and dependencies properly on an ubuntu system? im sure someone in the community more familiar with rhel suse etc can modify it to their distros somewhere along the line as well

**Comment by griffinwebnet** [comment]
i actually dont have an older version. it requires 3.3.1 im on 3.18.0. the latest Maven in the Ubuntu Repos

Correct me if im wrong but a 3.18 minor is newer than a 3.3?

**Comment by TylerJewell** [comment]
Yep, everything else you did was spot on.  You just got a bad compile error.

The latest maven in the ubuntu repo may not be latest enough for compiling the package.  You can download a new maven version from their ZIP file on the apache maven site.  It's just an extraction of a ZIP and then update your path to use the new binary.

In terms of the docs, we try to channel new users into one of two paths:
1. If you are just a user, then the binary install and then we take you to our usage pages (install and just run).
2. If you want to work off the source, we tend to keep you out of the docs initially and start you on the Che repo itself.  So this is the main repo README.md & CONTRIBUTING.md, which should provide very detailed instructions for how to compile, set error codes, debug, and make a pull request.  If there are ways we can improve this, definitely we love the contributions.

**Comment by griffinwebnet** [comment]
well let me see if i can get a newer maven. i may install the binaries to play around with but i wanted to go from source so i could tinker. My end goal is to set this us for my development team to use have have their projects. i also want to configure git to create a new repo for ever project's working folder (if this doesn't already happen with che) this way my guys who prefer their own editor can use it and still contribute to the cloud codebase for projects. plus i also wanted to be able to create users who cant create projects because i occasionally hire freelancers and i dont want them making projects on the server, or seeing projects they arent hired for. I havent found a platform yet that does all of this out of the box hence the source compilation here for tinkering purposes

**Comment by TylerJewell** [comment]
I like the spirit!

On the configuration, you have many choices for your team:
1. You can create a workspace for each team member
2. You could create multiple projects (each to their own repo) in a single workspace
3. You can checkout only a portion of a folder from a git repo
4. You can set up a workspace and then let your guys SSH into the workspace, using their desktop tools in the remote workspace.

All of the workspace & project management can be controlled over REST API, too.

**Comment by griffinwebnet** [comment]
oh neat! so i could for instance do a workspace called someAPP

then inside have a project for the nodejs backend, a project for the Angular dekstop/web frontend, and a project for the phonegap backend , and set it so certain users on the sever can only see they project in that workspace, OR see only that particular workspace?

This can be done without any tinkering, just from the stock binaries?

**Comment by TylerJewell** [comment]
You can do most of that with the binaries.
- Yes to provisioning a new workspace over API.
- Yes, that workspace can have a recipe which defines its runtime
- Yes, that workspace can have a template which auto-clones from a branch or ref-spec
- Yes, you can have many worskapces in a single server

But the system doesn't have resource limitations per workspace or security controls per user (in Che).  Che does have an embedded authentication and authorization system, but out of the box it is configured for a single identity shared by all worskpaces.  You have three choices to extend this:  (hard) you can implement new Che implementations of various APIs to provide multi-user identity, (easier) you can launch different che servers for each user with a router in front so that each user has their own private workspace server restricted and isolating resources and projects and workspaces, (commercial) codenvy is a multi-tenant, elastic implementation of che that applies user controls and permissions.  It's free for up to 5 users under fair source (fair.io) license.  So if you are a small shop, you can install it and get full control.

**Comment by griffinwebnet** [comment]
Ok so i just grabbed the binary and i am still receiving the java error so i will enabled the workaround we discussed previously.

i just saw that you asked earlier for the result of: `java -version 2>&1 | awk -F '"' '/version/ {print $2}'`

the output on my system is `1.8.0_72`

Perhaps the additional `0_72` is throwing the script off, but this is a typical output on Ubuntu, i cant speak for other versions.

Now it would seem che has started successfully, yet i cannot access it via web browser:

`che@codehost: ./che.sh -r:my.ip.add.res start`
`Docker is natively installed and reachable...`
`############## HOW TO CONNECT YOUR CHE CLIENT ###############`
`After Che server has booted, you can connect your clients by:`
`1. Open browser to http://localhost:8080, or:`
`2. Open native chromium app.`
`#############################################################`
`Using CATALINA_BASE:   /srv/eclipse-che-4.0.0-beta-13-1/tomcat`
`Using CATALINA_HOME:   /srv/eclipse-che-4.0.0-beta-13-1/tomcat`
`Using CATALINA_TMPDIR: /srv/eclipse-che-4.0.0-beta-13-1/tomcat/temp`
`Using JRE_HOME:        /usr/lib/java`
`Using CLASSPATH:       /srv/eclipse-che-4.0.0-beta-13-1/tomcat/conf/:/usr/lib/java/lib/tools.jar:/srv/eclipse-che-4.0.0-beta-13-1/tomcat/bin/bootstrap.jar:/srv/eclipse-che-4.0.0-beta-13-1/tomcat/bin/tomcat-juli.jar`
`Tomcat started.`

**Comment by TylerJewell** [comment]
I see that you added the -r:<ip-address> option on the command line.  As a first starting point, it would be easiest if you just run it as ./che.sh run so we can see all of the output.  Then we can get it working on localhost:8080, and then you can add in a remote IP address if your browser will not be on localhost itself.

**Comment by griffinwebnet** [comment]
Well this is a VPS with a CLI only interface so i wouldnt have any way to check if localhost is allowing connections to the IDE hence the -r but here is the output when running without the -r option:

`che@codehost: ./che.sh start`
`Docker is natively installed and reachable...`
`############## HOW TO CONNECT YOUR CHE CLIENT ###############`
`After Che server has booted, you can connect your clients by:`
`1. Open browser to http://localhost:8080, or:`
`2. Open native chromium app.`
`#############################################################`
`Using CATALINA_BASE:   /srv/eclipse-che-4.0.0-beta-13-1/tomcat`
`Using CATALINA_HOME:   /srv/eclipse-che-4.0.0-beta-13-1/tomcat`
`Using CATALINA_TMPDIR: /srv/eclipse-che-4.0.0-beta-13-1/tomcat/temp`
`Using JRE_HOME:        /usr/lib/java`
`Using CLASSPATH:       /srv/eclipse-che-4.0.0-beta-13-1/tomcat/conf/:/usr/lib/java/lib/tools.jar:/srv/eclipse-che-4.0.0-beta-13-1/tomcat/bin/bootstrap.jar:/srv/eclipse-che-4.0.0-beta-13-1/tomcat/bin/tomcat-juli.jar`
`Tomcat started.`

Looks to be the same output to me... output of -d option

`che@codehost: ./che.sh -d start`
`USE_DOCKER: false`
`CHE_DOCKER_TAG: latest`
`CHE_PORT: 8080`
`CHE_IP: ""`
`CHE_DOCKER_MACHINE: default`
`LAUNCH_REGISTRY: false`
`PRINT_CLIENT_CONNECT: true`
`CHECK_DOCKER_UID: true`
`USE_HELP: false`
`CHE_SERVER_ACTION: start`
`USE_DEBUG: true`
`Docker is natively installed and reachable...`
`############## HOW TO CONNECT YOUR CHE CLIENT ###############`
`After Che server has booted, you can connect your clients by:`
`1. Open browser to http://localhost:8080, or:`
`2. Open native chromium app.`
`#############################################################`
`Using CATALINA_BASE:   /srv/eclipse-che-4.0.0-beta-13-1/tomcat`
`Using CATALINA_HOME:   /srv/eclipse-che-4.0.0-beta-13-1/tomcat`
`Using CATALINA_TMPDIR: /srv/eclipse-che-4.0.0-beta-13-1/tomcat/temp`
`Using JRE_HOME:        /usr/lib/java`
`Using CLASSPATH:       /srv/eclipse-che-4.0.0-beta-13-1/tomcat/conf/:/usr/lib/java/lib/tools.jar:/srv/eclipse-che-4.0.0-beta-13-1/tomcat/bin/bootstrap.jar:/srv/eclipse-che-4.0.0-beta-13-1/tomcat/bin/tomcat-juli.jar`
`Tomcat started.`

A Note on what you said above:

i did check out codenvy the problem is i have more than 5 users just not at the same time. freelancers float in and out and all of my team is proficient on different aspect, not all of which are always needed all the time, therefore they float in and out of needing to use the IDE, so codenvy's free and paid plans didnt exactly work. plus i tried to sign up for the on prem free plan and they said they are in private alpha and that it wont be ready for a few more months and on the the SAAS plan its pretty easy to accidentally chew up 20GB Hours over there (i went through 8 of them by accidentally leaving an apache server running and closing the browser while demoing it out) BUT i really loved the editor which is what brought me here. ive seen you are the lead developer on this and i have to say you have seriously done some amazing work and im looking forward to playing around with it.

**Comment by TylerJewell** [comment]
The output that you got from the command is the right output.  Bash has operators that can read 1.8.0_72 and understand that 1.8 is less than that.  

Can you confirm that in the binary that you installed, in the che.sh that you ran, that the Java check has this syntax?  This was a slight change that was made after the beta, but before a bug fix release. I would like to ensure that it was incorporated into your scripts.  If this is the same and you still get the error then we'll need a live debug session.  We have tested extensively on Ubuntu, so something else is getting in the way.

```
JAVA_VERSION=$("${JAVA_HOME}/bin/java" -version 2>&1 | awk -F '"' '/version/ {print $2}')
if [[  -z "${JAVA_VERSION}" || "${JAVA_VERSION}" < "1.8" ]]; then
        error_exit "Che requires Java version 1.8 or higher. We found a lower version."
        return
 fi
```

**Comment by TylerJewell** [comment]
Ok, if it's a VPC, then you must do two things;
 -r:<ip-address-of-VPC>.  In the first run that you did, the client was saying "localhost" so you didn't give it the proper -r input.  Also run it with che.sh run, not start, otherwise we cannot see the tomcat out put and how it is binding itself.

so: ./che.sh --remote:192.168.99.100 run

**Comment by griffinwebnet** [comment]
ok so i initialy got an error thatr  /usr/lib/java/bin didnt exist so i  did export JAVA_HOME with the fill path to the java binary instead of the symbolic one and now it seems to work though the IDE is REALLY slow. almost 2 minutes of loading for a page to come up.

in regard to start vs run, you should modify that in the readme which denotes:

`# Executable files are:`
`bin/che.sh  [ start | stop ]`
`bin/che.bat [ start | stop ]`

**Comment by TylerJewell** [comment]
Yes, I think the readme can be clearer.  If you do not provide any option, then ./che.sh run is the default.  start and stop are other options.  We will update that.

The slow running seems quite odd - pages should load immediately.

**Comment by griffinwebnet** [comment]
once the IDE Is loaded its quite snappy but first load is a r ough one. let me try again in a new tab. perhaps i was trying to access it too quickly after startup.

Edit: Yup it is pretty quick now. Nest lets see if i can uncomment that java section in the starup script

**Comment by TylerJewell** [comment]
Well the first time you create a workspace, the workspace recipe references a Docker image that has to be downloaded from DockerHub.  And that image is about 200MB.  But the project creation screen gives you a status update on the progress there.  After the image is cached on your server, then all projects that go into workspaces based upon that image use the cached one.

**Comment by griffinwebnet** [comment]
Ahhhh hthat makes sense. so when the che.sh is starting its actually spawning a docker for that instance? or is it spawning a new docker for each workspace (this is the impression i was getting based on the resource allocation selector in workspace creation)

**Comment by TylerJewell** [comment]
Every workspace has at least one environment which is powered by a set of docker containers.  When you chose the stack that is the docker image definition to use to power the workspace.  We download that image and save it to the server.  All new workspaces that use that same stack will just instantiate from that image which is milliseconds.

**Comment by griffinwebnet** [comment]
ok so the last thing i cant figure out to start playing with this, is how do i get che to launch on system start? or without an active terminal?

**Comment by TylerJewell** [comment]
You can do a few things:
- First, running ./che.sh start will launch tomcat in the background
- Second, you can also run the Tomcat startup scripts directly, which provide more options.  These scripts are in /tomcat/bin.
- Third, you can then add Che startup as a daemon that is launchable in your linux configuration.

I am going to close off this issue for tracking as I believe the original issue has been resolved. You can file other issues if necessary.
